### PR TITLE
feat(rr): config parsing

### PR DESCRIPTION
**Pull request by seaerchin**
## Problem
Our current model of parsing is simplified and only works for pages; this PR adds parsing for configuration files (nav + setting) 

Closes [IS-54]

## Solution
1. add a config service that handles whether a file is a configuration file or not
2. extend dto definitions to differentiate between the types 
3. in the event of failure, return the filename + folder path to the frontend to display

[IS-54]: https://isomeropengov.atlassian.net/browse/IS-54?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/routes/v2/authenticated/sites.ts
Comment (seaerchin) [reply src/routes/v2/authenticated/sites.ts]: the type is `Result<T, E>` and the `err` here is a _variant_ of `Result` (either `ok` or `err`); the actual error itself is contained within `err`

File: src/services/fileServices/YmlFileServices/ConfigService.ts
Comment (harishv7) [inline src/services/fileServices/YmlFileServices/ConfigService.ts:40]: use enums

File: src/errors/PageParseError.ts
Comment (harishv7) [reply src/errors/PageParseError.ts:3]: So for example IsomerError(httpStatus, internalErrCode, errMsg)

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [inline src/services/review/ReviewRequestService.ts:143]: might want to abstract this as a type and use with a readable name

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [inline src/services/review/ReviewRequestService.ts:152]: why are we using snake_case here

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [inline src/services/review/ReviewRequestService.ts:11]: have we decided to use neverthrow across our codebase?

Comment (harishv7) [reply src/services/review/ReviewRequestService.ts:11]: I think is an interesting way to handle errors, but it's quite a shift in the way conventional errors are handled in js

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [inline src/services/review/ReviewRequestService.ts:219]: what are the types there are? use enum?

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [inline src/services/review/ReviewRequestService.ts:238]: use enums

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [inline src/services/review/ReviewRequestService.ts:177]: shld we do a length check first?

is there a possibility ref= doesn't exist? if so, that would be an error state right (thought it technically wont break the code)

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [inline src/services/review/ReviewRequestService.ts:189]: sorry can i check where does this unWrapOr come from? what does it do?

File: src/routes/v2/authenticated/review.ts
Comment (harishv7) [inline src/routes/v2/authenticated/review.ts:87]: i see a lot of these custom pipes - are these from neverthrow?

File: src/types/dto/review.ts
Comment (harishv7) [inline src/types/dto/review.ts:25]: using enums will be cleaner

File: src/services/fileServices/YmlFileServices/ConfigService.ts
Comment (harishv7) [inline src/services/fileServices/YmlFileServices/ConfigService.ts:13]: do we have a test file for this new service?

File: src/errors/ConfigParseError.ts
Comment (seaerchin) [reply src/errors/ConfigParseError.ts:5]: u r right; we should also do this in 1 shot... i'll change the errors introduced here but we fundamentally need to rehaul our error hierarchy + structure sadge

could i trouble you to file an issue on this (error overhaul) either on github or jira?

File: src/errors/PageParseError.ts
Comment (seaerchin) [reply src/errors/PageParseError.ts:3]: re: 1 error file per error - no strong opinions, style when i was introduced to the codebase was 1/file, which is inline w/ other codebases so i'm ok with it too. our errors are classes so eslint might complain if there's more than a certain # per file

re: error hierarchy - agreed; should relook at this later on.

File: src/services/fileServices/YmlFileServices/ConfigService.ts
Comment (seaerchin) [reply src/services/fileServices/YmlFileServices/ConfigService.ts:40]: i like unions more tbh, i've ran into type errors w/ enums in the past due to how they are implemented in TS and hence, i've avoided them like the plague ever since.

see [here](https://fettblog.eu/tidy-typescript-avoid-enums/) or [here](https://www.niraj.life/blog/avoid-using-enums-in-typescript) for more details  but if you feel strongly about this, i can change it

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:143]: caught me being lazy AHHAHHA wdyt about `shaMappings`

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:152]: github returns as snake_case, renamed to `contents`

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:11]: agreed, but throwing isn't good because then devs aren't inclined to handle errors properly + there's no oversight onto what errors are actually being returned. 

this forces errors to be explicit and handled properly rather than throwing and expecting a middleware to do so somewhere (as is done now)

File: src/routes/v2/authenticated/review.ts
Comment (seaerchin) [reply src/routes/v2/authenticated/review.ts:87]: yeap

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:189]: `neverthrow` - if it's ok, returns the value in ok, otherwise uses a default

File: src/types/dto/review.ts
Comment (seaerchin) [reply src/types/dto/review.ts:25]: see above ;--;

File: src/services/fileServices/YmlFileServices/ConfigService.ts
Comment (seaerchin) [reply src/services/fileServices/YmlFileServices/ConfigService.ts:13]: not at present, will add in a future PR

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:219]: see above ;--;

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:238]: see above ;--;

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:177]: re: this - the github schema [here](https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#compare-two-commits) has a `?ref=` at the end, which is why i made this assumption

i think it's possible that it doesn't exist but it does imply that github made a breaking change to their api

File: src/services/review/ReviewRequestService.ts
Comment (kishore03109) [inline src/services/review/ReviewRequestService.ts]: oof might be out of scope of this PR, but do you know what this todo is for?

File: src/utils/files.ts
Comment (kishore03109) [inline src/utils/files.ts:26]: I am confused here, why are we wrapping an err around an ok here ah?

File: src/services/fileServices/MdPageServices/PageService.ts
Comment (kishore03109) [inline src/services/fileServices/MdPageServices/PageService.ts:136]: Nit: can we just return a boolean value instead? Using neverthrow here feels too verbose for a function like this

File: src/services/fileServices/MdPageServices/PageService.ts
Comment (kishore03109) [inline src/services/fileServices/MdPageServices/PageService.ts:196]: Do we want to return the specific file that wasnt found here? (ie the path?)

File: src/services/fileServices/MdPageServices/PageService.ts
Comment (kishore03109) [inline src/services/fileServices/MdPageServices/PageService.ts:238]: May I get your help to add a comment of how a `rawPath` might look like? abit confused what the each element of `rawPath` represents

**Comment by kishore03109** [comment]
CICD tests are failing, may I trouble you you to take a look at them?

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts]: i don't see this specific `TODO` actually, you might have reviewed an old commit

File: src/services/fileServices/MdPageServices/PageService.ts
Comment (seaerchin) [reply src/services/fileServices/MdPageServices/PageService.ts:136]: actually the problem here is that we want to distinguish `MarkdownPages` from other things (eg: `YmlConfig` or `HtmlPages`) - the problem with returning `boolean` is that we lose context wrt what the `pageName` is beyond its immediate containing function.

i can change this to `string is MarkdownPage` and brand this string to prevent the verbosity - wdyt

File: src/services/fileServices/MdPageServices/PageService.ts
Comment (seaerchin) [reply src/services/fileServices/MdPageServices/PageService.ts:238]: can but i'll add it onto the `PathInfo` type - this is to avoid repetition + set at the containing type. wdyt?

Comment (seaerchin) [reply src/services/fileServices/MdPageServices/PageService.ts:238]: can but i'll add it onto the `PathInfo` type - this is to avoid repetition + set at the containing type. wdyt?

**Comment by seaerchin** [comment]
> CICD tests are failing, may I trouble you you to take a look at them?

kek i think the source branch is `feat/buttons`, which had tests fixed in #587 and #588; i'll probably rebase + merge those first and then rebase this on top once approved to see if the tests are still failing

File: src/errors/ConfigParseError.ts
Comment (harishv7) [reply src/errors/ConfigParseError.ts:5]: Sure, here: https://isomeropengov.atlassian.net/browse/IS-37

File: src/utils/files.ts
Comment (kishore03109) [reply src/utils/files.ts:26]: uhm just checking in, we expect the caller to meaningfully use the `name` then? if yes, feel free to resolve this.

File: src/services/fileServices/MdPageServices/PageService.ts
Comment (kishore03109) [reply src/services/fileServices/MdPageServices/PageService.ts:136]: hmm isnt that beyond the scope of this function, which is to just tell if this page is of type `MarkdownPages`? 
if not should we then rename this to `getPageType` right?

File: src/services/review/ReviewRequestService.ts
Comment (kishore03109) [reply src/services/review/ReviewRequestService.ts:177]: wait based on the schema, why are we not taking the `base_commit`.`sha` directly ah?

File: src/errors/PageParseError.ts
Comment (harishv7) [reply src/errors/PageParseError.ts:3]: Okay I think can keep like this, but we need a revamp of the error hierarchy soon

File: src/services/fileServices/YmlFileServices/ConfigService.ts
Comment (harishv7) [reply src/services/fileServices/YmlFileServices/ConfigService.ts:40]: @seaerchin That's a good point! I think I used to do it like this:

```
enum ESomeEnum {
    ValOne: "1",
    ValTwo: "2"
}

type EFinalType = ESomeEnum.ValOne | ESomeEnum.ValTwo
```

Then we use the EFinalType for the actual assertion

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [reply src/services/review/ReviewRequestService.ts:143]: Hahaha! 😂 

Shall we define one `shaMapping` and then `shaMappings` is a Record of string to `shaMapping`?

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [reply src/services/review/ReviewRequestService.ts:189]: Ooh okay, I think this syntax is going take some time to get used to. Do you think we need neverthrow though? I thought of discussing with you how best to standardise returns and responses across the codebase first

File: src/services/fileServices/YmlFileServices/ConfigService.ts
Comment (seaerchin) [reply src/services/fileServices/YmlFileServices/ConfigService.ts:40]: this actually results in the same thing HAHA

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:177]: sorry, where did you see this? i saw mine under `Comparing two commits`, where the schema explicitly has a `?ref=` at the end (returned by github)

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:189]: shure, can!

File: src/utils/files.ts
Comment (seaerchin) [reply src/utils/files.ts:26]: yea - if it's rooted at `/` (homepage is `/index.md` for eg), then the `name` will be `index.md` but path is empty (`err`).

hence, `name` is always semantically meaningful but path depends

File: src/errors/ConfigParseError.ts
Comment (seaerchin) [reply src/errors/ConfigParseError.ts:5]: marking as resolved cos we're tracking on jira

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:177]: @kishore03109 i've a comment there (up above) and it's cos the sha present is a checksum of the files' contents but we want the sha referencing the actual file

File: src/services/fileServices/YmlFileServices/ConfigService.ts
Comment (seaerchin) [reply src/services/fileServices/YmlFileServices/ConfigService.ts:40]: oh ps, i haven't had time to look at this due to being caught up in the release but this has the same effect because the discriminated union still only allows the same set of values.

File: src/services/review/ReviewRequestService.ts
Comment (seaerchin) [reply src/services/review/ReviewRequestService.ts:143]: `Record` takes 2 type arguments, so i think sha mapping would be misleading (cos that's the actual user info that we'd be interested in).

wdyt about 
```ts
type GithubEditInfo = {
        author: string
        unixTime: number
}
type Sha = string
type ShaMappings = Record<Sha, GithubEditInfo>
```

File: src/services/fileServices/YmlFileServices/ConfigService.ts
Comment (harishv7) [reply src/services/fileServices/YmlFileServices/ConfigService.ts:40]: yes true, but felt it is cleaner with the use of discriminated unions + enums rather than strings

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [reply src/services/review/ReviewRequestService.ts]: okay makes sense to me

File: src/services/review/ReviewRequestService.ts
Comment (harishv7) [reply src/services/review/ReviewRequestService.ts:11]: back to the discussion the team had on using neverthrow, I guess we will be postponing the use of the library till we get clarity on some of the refactorings we need to do in our code right?

File: src/types/pages.ts
Comment (kishore03109) [inline src/types/pages.ts:27]: sanity check: Is this removal of `kind` intended?

**Review by kishore03109 — APPROVED** [review]
[Tests] The full testing plan will be in the corresponding FE PR for buttons right? If not, please dismiss this review

File: src/types/pages.ts
Comment (seaerchin) [reply src/types/pages.ts:27]: ye shifted to util

File: src/routes/v2/authenticated/review.ts
Comment (seaerchin) [reply src/routes/v2/authenticated/review.ts:139]: the `NotFoundError` was constructed either with no arguments (defaults to generic error messages of `Something went wrong`) or a default string (no sensitive info) or a formatted string containing the requested `rawPath`, similarly for `MissingSiteError`, it's just constructed with a default message of `the requested site was not found in isomer`, which i think is ok.

lmk if `rawPath` is a concern

File: src/integration/Reviews.spec.ts
Comment (seaerchin) [reply src/integration/Reviews.spec.ts:274]: can i clarify this - isn't the removed property added back below?

File: src/integration/Reviews.spec.ts
Comment (seaerchin) [reply src/integration/Reviews.spec.ts:758]: can i clarify this - isn't the removed property added back below?

**Pull request merged**
